### ISSUE TITLE
Another "Input buffet contains unsupported image format"

### ISSUE THREAD

**Issue by JakobJingleheimer** [lovell/sharp#2681]
> Are you using the latest version?

Yes: `0.28.1`

> What are the steps to reproduce?

1. Use `fs.watch` to observe a (USB-3) flash drive
1. On `"rename"`, call the below (where `pathToFile` is like `'/Volumes/FlashDrive/_MG_7743.JPG'`)
1. Drag `_MG_7743.JPG` onto the flash drive.

```js
const file = await fs.promises.readFile(pathToFile);

// …

sharp(file)
  .resize({ width: 800 })
  .toFile(dest);
```

> What is the expected behaviour?

For most images, this works fine. However, for seemingly random images (which otherwise appear to be fine), the Input Buffer unsupported image format is consistently thrown. I took all the images myself with the same camera, so I'm certain they're actually jpeg (unlike #1852, where the image was secretly a bitmap in disguise).

Curiously, the error is NOT thrown without `FSWatcher` in the mix (eg when the image is already on the flash drive).

Also curiously, the error is NOT thrown when `fs.watch` observes a same-disk-to-same-disk move (eg moving the image from elsewhere on the local hard drive to the watched folder on the local hard drive, or moving the image from elsewhere on the flash drive to the watched folder on the hard drive). That makes me think it's some kind of latency issue, BUT:

The `file` buffer appears to be intact in both scenarios, and the `readFile` is `await`ed, so it should have finished before the buffer is passed to Sharp.

However, in a case against latency (in addition to it being always the same images): It does not appear to be affected by file-size: some of the erroring ones are significantly smaller than others that non-error (ex one erroring is 2.5 MB, and one non-erroring is 4.5 MB). If it were a latency issue, larger files should be more prone to errorure than smaller (out of my test set of 18 images, 8 error, and their sizes are roughly evenly distributed in a range of 2.5 MB to 4.5 MB, and the 10 non-erroring are similarly ranged in size).

And finally, the error does not actually seem to matter—the operation still succeeds: the 800-width image gets written to `dest`, and AFAICT, the new file is fine.

My apologies if the problem is actually unrelated to Sharp (that seems possible since `fs.watch` appears to be a critical piece, but that doesn't make much sense either).

> Are you able to provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem?

Yes, I can put one together if needed.

> Are you able to provide a sample image that helps explain the problem?

Yes, can do.

> What is the output of running `npx envinfo --binaries --system`?

```
  System:
    OS: macOS 10.15.7
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 15.41 GB / 32.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 15.14.0 - ~/.nvm/versions/node/v15.14.0/bin/node
    npm: 7.7.6 - ~/.nvm/versions/node/v15.14.0/bin/npm
```

**Comment by lovell** [comment]
This almost certainly relates to data on the USB device not being completely written before the event fires, or possibly the event firing multiple times.

A "rename" involving two physical devices will write the new file then delete the old file, rather than e.g. just updating inode metadata.

Please try [chokidar](https://github.com/paulmillr/chokidar) and make sure you set the `awaitWriteFinish` option to `true`.

**Comment by JakobJingleheimer** [comment]
Okay, I shall give that a try.

Any idea why it's always the same files that error and don't error though? That seems pretty suspicious to me. I tried a couple dozen times, and then also with a much larger set of files (~700) twice.

When dragging the files onto the flash drive, a copy is made (the originals are not deleted); but when moving on the same disk, the original is gone (I'm not sure whether it is a deletion of the 'original' or a pointer update—I would _hope_ the latter).

Thanks!

**Comment by JakobJingleheimer** [comment]
I added some logging, and it appears that `FSWatcher` fires 2 `rename` events:

```
Watcher: { eventType: 'rename', filename: '_MG_7446.JPG' }

Processing: _MG_7446.JPG

Error: Input buffer contains unsupported image format

Watcher: { eventType: 'rename', filename: '_MG_7446.JPG' }

Processing: _MG_7446.JPG
```

```
Watcher: { eventType: 'rename', filename: '_MG_7442.JPG' }

Processing: _MG_7442.JPG

Watcher: { eventType: 'rename', filename: '_MG_7442.JPG' }

Processing: _MG_7442.JPG
```

* _MG_7442 never errors and is 4.5 MB
* _MG_7446 always errors and is 2.5 MB

Sequence doesn't matter: _MG_7442 before or after _MG_7446 doesn't change erroring vs not erroring, and one-at-a-time also always produces the same result.

**Comment by lovell** [comment]
Thanks for the update, glad you worked it out.

**Comment by JakobJingleheimer** [comment]
Hmm? I didn't work it out.

I think this essentially proves it's not latency.

**Comment by lovell** [comment]
Oh, I thought your comment meant that you'd worked out the problem relates to race conditions within `fs.watch` and is unrelated to sharp.

**Comment by JakobJingleheimer** [comment]
It doesn't appear to be a race condition, but noticing the duplicate events did put me onto a simple workaround (chokidar may also handle), which I just verified works: keep a list of files signalled by FSWatcher and use that to always ignore the 1st event (since the 2nd is always safe but the 1st isn't).

Fortunately, I already keep a similar list elsewhere as part of the processing (to avoid truly duplicate files), so I just had to move where/when that happens.

So in summary, I don't know what the actual underlying problem is, but I did also confirm that the file buffer passed to sharp the first time (when sharp throws) is indeed fine: I instead immediately wrote the buffer to disk as a file, and upon viewing the image (with Preview), I cannot find any issues (it renders, no errors, EXIF data looks fine, etc).

Thank you for the help!

**Comment by JakobJingleheimer** [comment]
@lovell I did some more digging, and it seems in the above scenario when sharp is checking the buffer, the buffer is empty. Would you consider adding a check and error message specifically about that? 

```
if (!buffer.length) throw new Error('Input buffer is empty');
```

I think that would be much more helpful than saying it contains an unsupported image (which is technically not accurate: it doesn't contain anything 😜).

**Comment by lovell** [comment]
Yes please, a PR to throw an `Error` when attempting to create a `sharp` instance with a zero-length `Buffer` or `Uint8Array` would be most welcome.

**Comment by JakobJingleheimer** [comment]
@lovell I'm looking for the appropriate place to put that check. [input.js → _createInputDescriptor()](https://github.com/lovell/sharp/blob/master/lib/input.js#L22) looks good. Thoughts?

**Comment by lovell** [comment]
Yes, `_createInputDescriptor` would be a suitable place for this check, plus `test/unit/raw.js` for unit tests.

**Comment by lovell** [comment]
(There are some raw input tests in `test/unit/io.js` but they should probably be moved to `test/unit/raw.js`.)
